internal/controller/gateway.go

```diff
@@ -31,46 +31,107 @@ import (
+	// referencedClass is the gatewayclass referenced by managed gateways.
+	referencedClass *gatewayapi_v1alpha1.GatewayClass
 }
 
 // NewGatewayController creates the gateway controller from mgr. The controller will be pre-configured
-// to watch for Gateway objects across all namespaces.
-func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+// to watch for Gateway objects across all namespaces and reconcile those that match class.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger, class string) (controller.Controller, error) {
 	r := &gatewayReconciler{
-		client:       mgr.GetClient(),
-		eventHandler: eventHandler,
-		FieldLogger:  log,
+		ctx:            context.Background(),
+		client:         mgr.GetClient(),
+		eventHandler:   eventHandler,
+		log:            log,
+		controllerName: class,
 	}
 	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return nil, err
 	}
-	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, r.enqueueRequestForOwnedGateway()); err != nil {
 		return nil, err
 	}
 	return c, nil
 }
 
+// enqueueRequestForOwnedGateway returns an event handler that maps events to
+// Gateway objects that reference a GatewayClass owned by Contour.
+func (r *gatewayReconciler) enqueueRequestForOwnedGateway() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		gw, ok := a.(*gatewayapi_v1alpha1.Gateway)
+		if !ok {
+			r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+				Info("invalid object, bypassing reconciliation.")
+			return []reconcile.Request{}
+		}
+		gc, err := r.classForGateway(gw)
+		if err != nil {
+			r.log.WithField("namespace", gw.Namespace).WithField("name", gw.Name).Error(err)
+			return []reconcile.Request{}
+		}
+		r.referencedClass = gc
+		return []reconcile.Request{
+			{
+				NamespacedName: types.NamespacedName{
+					Namespace: gw.Namespace,
+					Name:      gw.Name,
+				},
+			},
+		}
+	})
+}
+
+// classForGateway returns nil, error if the gatewayclass referenced by gw does not exist
+// or is not owned by Contour. Otherwise, the referenced gatewayclass is returned.
+func (r *gatewayReconciler) classForGateway(gw *gatewayapi_v1alpha1.Gateway) (*gatewayapi_v1alpha1.GatewayClass, error) {
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(r.ctx, types.NamespacedName{Name: gw.Spec.GatewayClassName}, gc); err != nil {
+		return nil, fmt.Errorf("failed to get gatewayclass %s: %w", gw.Spec.GatewayClassName, err)
+	}
+	if !r.matchesControllerName(gc) {
+		return nil, fmt.Errorf("gatewayclass %q not owned by contour", gc.Name)
+	}
+	return gc, nil
+}
```

Comment:
This helper function (`classForGateway`) does more than its name implies. I would maybe move the logic that checks if the controller name matches Contour's out of this helper and into the place that calls it, or otherwise perhaps change this helper to be `hasContourOwnedClass` and return a bool, since it doesn't look like the GatewayClass itself is used, we just need to know if it's Contour-owned. I think either of those would be clearer.